### PR TITLE
Authentication functionality via saml

### PR DESCRIPTION
**Pull request by Baggerone**

### CODE REVIEW

File: application/actions/auth.go [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/silinternational/riskman-api/auth"
```
Comment (Schparky) [inline application/actions/auth.go:11]: looks like `goimports` could clean this up a bit, removing empty lines and grouping `silinternational` imports

File: application/actions/auth.go [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+	ReturnToParam      = "return-to"
+	ReturnToSessionKey = "ReturnTo"
+
+	// http param for token type
+	TokenTypeParam = "token-type"
+)
+
+func replaceNewLines(input string) string {
+	output := strings.Replace(input, `\n`, "\n", -1)
+	return output
```
Comment (Schparky) [inline application/actions/auth.go:44]: reduce to 1 line?

File: application/actions/auth.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+	AssertionConsumerServiceURL: domain.Env.SamlAssertionConsumerServiceUrl,
+	IDPPublicCert:               replaceNewLines(domain.Env.SamlIdpCert),
+	SPPublicCert:                replaceNewLines(domain.Env.SamlSpCert),
+	SPPrivateKey:                replaceNewLines(domain.Env.SamlSpPrivateKey),
+	SignRequest:                 domain.Env.SamlSignRequest,
+	CheckResponseSigning:        domain.Env.SamlCheckResponseSigning,
+	AttributeMap:                nil,
+}
+
+type authError struct {
```
Comment (Schparky) [inline application/actions/auth.go:62]: We didn't use this in TF, can we use AppError here too?

File: application/actions/auth.go [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+	httpStatus int
+	errorKey   api.ErrorKey
+	errorMsg   string
+}
+
+// Make extras variadic, so that it can be omitted from the params
```
Comment (Schparky) [inline application/actions/auth.go:68]: remove `extras` parameter

File: application/actions/auth.go [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+	authUser.AccessToken = accessToken
+	authUser.AccessTokenExpiresAt = expiresAt
+
+	// set person on rollbar session
+	domain.RollbarSetPerson(c, user.StaffID, user.FirstName, user.LastName, user.Email)
+
+	return c.Redirect(302, getLoginSuccessRedirectURL(*authUser, returnTo))
+}
+
+// Make extras variadic, so that it can be omitted from the params
```
Comment (Schparky) [inline application/actions/auth.go:245]:
```suggestion
```

File: application/actions/auth.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+	c.Session().Clear()
+
+	return c.Render(authErr.httpStatus, render.JSON(appErr))
+}
+
+func setCurrentUser(next buffalo.Handler) buffalo.Handler {
+	return func(c buffalo.Context) error {
+		bearerToken := domain.GetBearerTokenFromRequest(c.Request())
+		if bearerToken == "" {
+			return c.Error(http.StatusUnauthorized, errors.New("no Bearer token provided"))
```
Comment (Schparky) [inline application/actions/auth.go:86]: are we going to use `ReportError` in this project?

File: application/models/user.go [hunk 7]
```diff
@@ -62,3 +71,48 @@ func (u *User) IsActorAllowedTo(tx *pop.Connection, actor User, p Permission, su
...
+			"cannot create token with empty clientID for user %s %s", u.FirstName, u.LastName)
+	}
+
+	token, _ := getRandomToken()
+	hash := HashClientIdAccessToken(clientID + token)
+	expireAt := createAccessTokenExpiry()
+
+	userAccessToken := &UserAccessToken{
+		UserID:      u.ID,
+		AccessToken: hash,
```
Comment (Schparky) [inline application/models/user.go:109]: Does the hash go into `TokenHash`? It seems we have a mix of implementation details pulled from WCA and TF, but I can't remember the exact differences.

**Review by Schparky — APPROVED** [review]

File: application/actions/auth.go (hunk 1 above)
Comment (Baggerone) [reply application/actions/auth.go:11]: It's not changing anything for me here.

Comment (Schparky) [reply application/actions/auth.go:11]: I guess it doesn't always regroup them if they're already separated.

File: application/actions/auth.go (hunk 2 above)
Comment (Baggerone) [reply application/actions/auth.go:44]: :+1:

File: application/actions/auth.go (hunk 3 above)
Comment (Baggerone) [reply application/actions/auth.go:62]: :+1:

File: application/actions/auth.go (hunk 4 above)
Comment (Baggerone) [reply application/actions/auth.go:68]: :+1:

File: application/actions/auth.go (hunk 5 above)
Comment (Baggerone) [reply application/actions/auth.go:245]: :+1:

File: application/auth/saml/provider.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	return rsaKey, []byte(c.SPPublicCert), nil
+}
+
+func New(config Config) (*Provider, error) {
+	p := &Provider{
+		Config: config,
+	}
+
+	err := p.initSAMLServiceProvider()
+	if err != nil {
```
Comment (jason-jackson) [inline application/auth/saml/provider.go:58]: Could just `return p, err`, no if statement needed

File: application/auth/saml/provider.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+func getSAMLAttributeFirstValue(attrName string, attributes []types.Attribute) string {
+	for _, attr := range attributes {
+		if attr.Name != attrName {
+			continue
+		}
+
+		if len(attr.Values) > 0 {
+			return attr.Values[0].Value
+		}
+		return ""
```
Comment (jason-jackson) [inline application/auth/saml/provider.go:150]: Alternatively `break` here for one non-existent exit point.

File: application/domain/domain.go [hunk 10]
```diff
@@ -202,3 +219,47 @@ func GetBearerTokenFromRequest(r *http.Request) string {
 
 	return string(matches[1])
 }
+
+// IsOtherThanNoRows returns false if the error is nil or is just reporting that there
+//   were no rows in the result set for a sql query.
+func IsOtherThanNoRows(err error) bool {
```
Comment (jason-jackson) [inline application/domain/domain.go:225]: If desired, could condense into one line, but I think what you have is cleaner.

`return err != nil && !strings.Contains(err.Error(), sql.ErrNoRows.Error()))`

File: application/domain/domain.go [hunk 11]
```diff
@@ -202,3 +219,47 @@ func GetBearerTokenFromRequest(r *http.Request) string {
...
+	if strings.Contains(err.Error(), sql.ErrNoRows.Error()) {
+		return false
+	}
+
+	return true
+}
+
+// RollbarSetPerson sets person on the rollbar context for further logging
+func RollbarSetPerson(c buffalo.Context, id, userFirst, userLast, email string) {
+
```
Comment (jason-jackson) [inline application/domain/domain.go:239]: extra whitespace

File: application/domain/domain.go [hunk 12]
```diff
@@ -202,3 +219,47 @@ func GetBearerTokenFromRequest(r *http.Request) string {
...
+	username := strings.TrimSpace(userFirst + " " + userLast)
+	rc, ok := c.Value(ContextKeyRollbar).(*rollbar.Client)
+	if ok {
+		rc.SetPerson(id, username, email)
+		return
```
Comment (jason-jackson) [inline application/domain/domain.go:244]: Any reason for a return here?

File: application/domain/warnerror.go [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	return string(j)
+}
+
+// rollbarMessage is a wrapper function to call rollbar's client.MessageWithExtras function from client stored in context
+func rollbarMessage(c buffalo.Context, level string, msg string, extras map[string]interface{}) {
+	rc, ok := c.Value(ContextKeyRollbar).(*rollbar.Client)
+	if ok {
+		rc.MessageWithExtras(level, msg, extras)
+		return
```
Comment (jason-jackson) [inline application/domain/warnerror.go:104]: Is return needed?

**Review by Schparky — APPROVED** [review]

File: application/actions/auth.go [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+}
+
+func setCurrentUser(next buffalo.Handler) buffalo.Handler {
+	return func(c buffalo.Context) error {
+		bearerToken := domain.GetBearerTokenFromRequest(c.Request())
+		if bearerToken == "" {
+			return reportError(c, &api.AppError{
+				HttpStatus: http.StatusUnauthorized,
+				Key:        api.ErrorNotAuthorized,
+				Message:    "no Bearer token provided",
```
Comment (jason-jackson) [inline application/actions/auth.go:67]: Bearer here is capitalized, all other places it is lowercase

File: application/actions/actions.go [hunk 15]
```diff
@@ -1 +1,110 @@
...
+
+	appErr.LoadTranslatedMessage(c)
+
+	// clear out debugging info if not in development or test
+	if domain.Env.GoEnv != "development" && domain.Env.GoEnv != "test" {
+		appErr.Extras = map[string]interface{}{}
+		appErr.DebugMsg = ""
+	}
+
+	if appErr.HttpStatus >= 300 && appErr.HttpStatus < 399 {
```
[the remark this answers is not in the document]
Comment (Baggerone) [reply application/actions/actions.go:43]: :+1:

File: application/auth/saml/provider.go (hunk 8 above)
Comment (Baggerone) [reply application/auth/saml/provider.go:58]: This is a borrowed-and-tweaked file. So, I'm trying to make only necessary changes, for next time that library changes the file and we have to find the differences.

File: application/domain/domain.go (hunk 12 above)
Comment (Baggerone) [reply application/domain/domain.go:244]: :+1:

File: application/domain/warnerror.go (hunk 13 above)
Comment (Baggerone) [reply application/domain/warnerror.go:104]: :+1:

File: application/actions/auth.go (hunk 14 above)
Comment (Baggerone) [reply application/actions/auth.go:67]: :+1:

**Pull request merged**